Move localizeUrl to new i18n-utils package

lsl opened the pull request:
#### Changes proposed in this Pull Request
* Moves lib/i18n-utils' `localizeUrl` function to a new package in `packages/i18n-utils`

#### Testing instructions

localizeUrl tests were included in the move:

```
yarn test-client client/lib/i18n-utils
yarn workspace @automattic/i18n-utils test
```

Manual testing involves checking `localizeUrl` usage still works as expected for example:

- Set locale to german or french using the account settings page http://calypso.localhost:3000/me/account
- Navigate to http://calypso.localhost:3000/me/privacy
- The link to `https://automattic.com/privacy` should be localized (e.g. `https://automattic.com/fr/privacy`) but should point to the original english url for any other language.

---

- Partially fixes #37665 - introduces / exposes localizeUrl for use in apps/packages.
- Enables a fix for #45610

scinos commented on client/lib/wporg/index.js:
```diff
@@ -37,6 +37,7 @@ function getWporgLocaleCode() {
 
 async function pluginRequest( url, body ) {
 	try {
+		// eslint-disable-next-line no-undef -- Probably: https://developer.mozilla.org/en-US/docs/Web/API/Fetch_API
```
It would be better to add `env: { browser: true }` to `./client/.eslintrc.js`. Does that removes the need to add this exclusion?

lsl replied:
[`ac46ab7` (#46709)](https://github.com/Automattic/wp-calypso/pull/46709/commits/ac46ab7f2aa93858ab98ee2043081260e4087ce8)

scinos commented on packages/i18n-utils/jest.config.js:
```diff
@@ -0,0 +1,5 @@
+module.exports = {
+	preset: '../../test/packages/jest-preset.js',
+	rootDir: __dirname,
```
I think you don't need `rootDir` or `cacheDirectory`, they come from the `preset`

scinos commented on packages/languages/bin/download.js:
```diff
@@ -1,24 +1,21 @@
 
 const LANGUAGES_META_URL = 'https://widgets.wp.com/languages/calypso/languages-meta.json';
-const FILE_PATH = 'client/languages/languages-meta.json';
+const FILE_PATH = path.resolve( __dirname, '../src/languages-meta.json' );
```
`path.join()` will give us the same result and is a bit more correct IMO

lsl replied:
[`56f0d23` (#46709)](https://github.com/Automattic/wp-calypso/pull/46709/commits/56f0d23af66b41528e943c53996fd010e6ec715f)

scinos commented on packages/languages/bin/download.js:
```diff
@@ -1,24 +1,21 @@
+/* eslint-disable no-console, no-process-exit, import/no-nodejs-modules */
```
This could be moved up to `./bin/.eslintrc.js`, and I think they will apply to any script inside `./bin`

lsl replied:
[`f50407a` (#46709)](https://github.com/Automattic/wp-calypso/pull/46709/commits/f50407add9123d769fc390f8c6c274e9bf7d473c)

scinos commented on packages/languages/jest.config.js:
```diff
@@ -0,0 +1,5 @@
+module.exports = {
```
I think you don't need `rootDir` or `cacheDirectory`, they come from the `preset`

lsl replied:
[`4ea1e04` (#46709)](https://github.com/Automattic/wp-calypso/pull/46709/commits/4ea1e0437b0dc126993ef86634c563a553d083a3)

lsl commented:
@yuliyan / @scinos 

Parts of this `i18n-utils` package make more sense to be part of `packages/languages`: e.g. `getLanguage( langSlug )` just performs a lookup on the language data with a fallback to parent locale if available, and `isTranslatedIncompletely` just returns a value of the language looked up by `getLanguage`

The remaining parts make more sense to be part of `package/i18n-calypso`. e.g. `canBeTranslated`, `translationExists`, `isMagnificentLocale` are all very a8c specific or depend on i18n-calypso to get the current locale (like `localizeUrl`)

My thinking is maybe to split i18n-utils up between the two existing packages and drop the notion of a new one. Does this sound ok **or is there still a push to drop `i18n-calypso` in favor of `@wordpress/i18n`?** In which case it may be wise to keep i18n-utils in play with some/none of it moved into languages. It would still depend on i18n-calypso but that can be dropped as soon as there is a replacement solution for getLocaleSlug (the only problematic dep.)

lsl commented:
@scinos I copy and pasted most of these config files / settings from elsewhere and have only looked into a handful of what each thing does so I very much appreciate the audit! :beers: 

Most of your comments were on code from the other PR #46709 so I've fixed there and rebased here. Thanks!

yuliyan commented:
> Parts of this `i18n-utils` package make more sense to be part of `packages/languages`: e.g. `getLanguage( langSlug )` just performs a lookup on the language data with a fallback to parent locale if available, and `isTranslatedIncompletely` just returns a value of the language looked up by `getLanguage`

That makes sense and I think moving them into `@automattic/languages` is fine. As for the remaining helpers, I like the idea of having them separate from `i18n-calypso` better. With `@wordpress/i18n` we would be forced to have them in a separate package anyway, so I think it will make the transition more smooth.

p-jackson commented on packages/languages/src/index.ts:
```diff
@@ -25,6 +26,7 @@ type SubLanguage = BaseLanguage & { parentLangSlug: string };
-const languages: Language[] = values( data );
-
+export const languagesBySlug = data;
+export const languageSlugs: LanguageSlug[] = keys( data );
```
You could always use `Object.keys( data )` instead of lodash. Doesn't make a big difference in the grand scheme of things, but it is one less dependency on a non-standard library.

lsl replied:
Yea, was just sticking with lodash because the values call did the same, probably should be Object.values / Object.keys. Is this a safe change or does lodash do something extra?

jsnajdr replied:
The Lodash versions accept a `null` or `undefined` argument and return an empty array for them. The standard `Object.*` versions will throw a `TypeError: Cannot convert undefined or null to object`.

Without Lodash, you need to check if the inputs are always valid objects/arrays and add some guards otherwise.

lsl replied:
So definitely safe now we're not downloading the file in the build. Will swap to those! thanks!

p-jackson reviewed (commented):
Looks good so far. I realise it's still a WIP, just left a couple notes. Let me know if there's anything I can help with.

lsl committed:
Add packages/i18n-utils

p-jackson committed:
Make localizeUrl's `toLocale` param optional

p-jackson committed:
Silenced unnecessary lint error in readme

p-jackson reviewed (commented):
I haven't quite finished my review yet: I can't seem to build calypso locally, and it also looks like there's a change where the new package isn't doing some sort of lazy load of something out of the config and I want to get to the bottom of that.

ramonjd commented on packages/i18n-utils/test/localize-url.js:
```diff
@@ -0,0 +1,333 @@
+		// Finally make sure that no overriding has stuck and it uses the getLocaleSlug() when no override is specified.
+		getLocaleSlug.mockImplementationOnce( () => 'ru' );
+		expect( localizeUrl( 'https://automattic.com/cookies/' ) ).toEqual(
+			'https://automattic.com/cookies/'
```
> Or maybe Russian isn't a mag16 language

Officially it is on WPCOM

Here's the list:

```
			'es',
			'pt-br',
			'de',
			'fr',
			'he',
			'ja',
			'it',
			'nl',
			'ru',
			'tr',
			'id',
			'zh-cn',
			'zh-tw',
			'ko',
			'ar',
			'sv',
```

lsl replied:
automattic.com cookie and privacy policies are translated with a manual process (needs legal input irrc), we only have translations for a couple of locales on these pages and defer to english in most situations.

I just copy pasted these tests from lib/i18n-utils, the test description should probably explain why the locale needs to be overridden but I suspect most of these cases are ported from the php version of this function.

lsl committed:
Update packages/i18n-utils/src/locales.ts

Co-authored-by: Philip Jackson <p-jackson@live.com>

lsl commented on packages/i18n-utils/test/localize-url.js:
```diff
@@ -0,0 +1,333 @@
+			( fullUrl ) => {
+				getLocaleSlug.mockImplementationOnce( () => 'en' );
+				expect( localizeUrl( fullUrl ) ).toEqual( fullUrl );
+				getLocaleSlug(); // make sure to consume it.
```
Lets leave as is so this file continues to match the client/lib version... Actually maybe its a good idea to drop that duplicate test code right now?

lsl replied:
further to that... all of these tests should just pass in the locale it wants to test, the mocking of getLocaleSlug should only really happen once to confirm the implementation uses getLocaleSlug if its not provided with a locale.

p-jackson replied:
Yes actually! We only have to test it the hard way once 👍

lsl commented on packages/i18n-utils/test/localize-url.js:
```diff
@@ -0,0 +1,333 @@
+			'https://automattic.com/de/cookies/'
+		);
+		getLocaleSlug.mockImplementationOnce( () => 'de' );
+		expect( localizeUrl( 'https://automattic.com/cookies' ) ).toEqual(
+			'https://automattic.com/de/cookies/'
+		);
```
It's probably an artifact of serverside version. We use a permalink structure that specifies trailing slash on urls in at least some if not all wp properties, so we probably have something that cleans these up in a filter running somewhere. I only learned about this recently so don't quote me as it might be off on some detail.

lsl committed:
Remove now duplicated tests

wp-desktop reviewed (dismissed):
WordPress Desktop CI Failure for job "wp-desktop-mac".

@lsl please inspect this job's build steps for breaking changes at [this link](https://circleci.com/gh/Automattic/wp-calypso/947082). For temporal failures, you may try to "Rerun Workflow from Failed".

Please also ensure this branch is rebased off latest Calypso.

p-jackson reviewed (approved):
Looks good 👍

lsl committed:
Bugfix + testcoverage

lsl commented:
This sort of non-side-by-side refactor is quite hard to do safely, caught a sloppy copy paste bug testing random links in the UI: [`59dd2bc` (#46998)](https://github.com/Automattic/wp-calypso/pull/46998/commits/59dd2bc43d7cfc5f574663361faf70913360041f)

yuliyan commented:
Are [#localizeUrl unit tests in lib/i18n-utils](https://github.com/Automattic/wp-calypso/blob/master/client/lib/i18n-utils/test/utils.js) now redundant? If so, let's remove them in a follow-up PR before they start failing.
